ratelimit/cluster.go

```diff
@@ -0,0 +1,136 @@
+package ratelimit
+
+import (
+	"math"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+	circularbuffer "github.com/szuecs/rate-limit-buffer"
+)
+
+// Swarmer interface defines the requirement for a Swarm, for use as
+// an exchange method for cluster ratelimits:
+// ratelimit.ClusterServiceRatelimit and
+// ratelimit.ClusterClientRatelimit.
+type Swarmer interface {
+	// ShareValue is used to share the local information with its peers.
+	ShareValue(string, interface{}) error
+	// Values is used to get global information about current rates.
+	Values(string) map[string]interface{}
+}
+
+// ClusterLimit stores all data required for the cluster ratelimit.
+type ClusterLimit struct {
+	name    string
+	local   limiter
+	maxHits int
+	window  time.Duration
+	swarm   Swarmer
+	resize  chan resizeLimit
+	quit    chan struct{}
+}
+
+type resizeLimit struct {
+	s string
+	n int
+}
+
+// NewClusterRateLimiter creates a new ClusterLimit for given Settings
+// and use the given Swarmer. Name is used in log messages to identify
+// the ratelimit instance.
+func NewClusterRateLimiter(s Settings, sw Swarmer, name string) *ClusterLimit {
+	rl := &ClusterLimit{
+		name:    name,
+		swarm:   sw,
+		maxHits: s.MaxHits,
+		window:  s.TimeWindow,
+		resize:  make(chan resizeLimit),
+		quit:    make(chan struct{}),
+	}
+	if s.CleanInterval == 0 {
+		log.Infof("new backend clusterRateLimiter")
+		rl.local = circularbuffer.NewRateLimiter(s.MaxHits, s.TimeWindow)
+	} else {
+		log.Infof("new client clusterRateLimiter")
+		rl.local = circularbuffer.NewClientRateLimiter(s.MaxHits, s.TimeWindow, s.CleanInterval)
+	}
+
+	// TODO(sszuecs): we might want to have one goroutine for all of these
+	go func() {
+		for {
+			select {
+			case size := <-rl.resize:
+				log.Debugf("resize clusterRatelimit: %v", size)
+				// TODO(sszuecs): call with "go" ?
+				rl.Resize(size.s, rl.maxHits/size.n)
```

Comment:
I think it was, because of not having Resize() in the hot path of the request.
Why does a request needs to wait for the resize of a datastructure?

Reply:
i see. Then let's keep it, and measure the effect as a follow up.

The reason why I am still not sure if we need it is that we normally have more pending requests than CPUs. So if we continue the goroutine of the request without waiting for the resize to finish, my asumption is, the entire throughput of requests won't be higher, because we are calling the resize on every single request, and so there's still the same amount of tasks to complete. Actually, it is even possible that the total throughput can get a little lower because of the higher number synchronization steps. Just guessing.

Both the current solution and the proposed change are optimizations and need proof, while the logic looks correct to me, so let's keep it now, and measure separately.